Question 1: What are the key differences between SQL and NoSQL databases?

Answer: SQL uses structured tables and fixed schemas, while NoSQL supports flexible schema.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

MessageError: Error: credential propagation was unsuccessful

Question 2: What makes MongoDB a good choice for modern applications?

Answer: MongoDB is flexible, scalable, and handles unstructured data.

Question 3: Explain the concept of collections in MongoDB.
Answer: A collection is a group of documents similar to tables.

Question 4: How does MongoDB ensure high availability using replication?

Answer: Through replica sets with primary and secondary nodes.

Question 5: What are the main benefits of MongoDB Atlas?
Answer: Managed cloud DB, backups, monitoring, scaling.

Question 6: What is the role of indexes in MongoDB?

Answer: They improve query performance.

Question 7: Describe stages of aggregation pipeline.

Answer: $match, $group, $project, $sort, $lookup.

Question 8: What is sharding in MongoDB?
Answer: Horizontal data distribution across servers.

Question 9: What is PyMongo?

Answer: Python driver for MongoDB.

Question 10: ACID properties in MongoDB.
Answer: Atomicity, Consistency, Isolation, Durability.

Question 11: Purpose of explain().

Answer: Shows query execution plan.

Question 12: Schema validation.
Answer: Done using JSON schema.

Question 13: Primary vs Secondary.

Answer: Primary writes, secondary replicates.

Question 14: Security mechanisms.

Answer: Auth, TLS, encryption.

Question 15: Embedded documents.

Answer: Related data inside single doc.

Question 16: Purpose of $lookup.

Answer: Join between collections.

Question 17: Common use cases.

Answer: Real-time apps, IoT, analytics.

Question 18: Horizontal scaling.

Answer: Increases capacity.

Question 19: Transactions difference.

Answer: MongoDB supports multi-doc transactions.

Question 20: Capped collections.

Answer: Fixed-size collections.

Question 21: Purpose of $match.

Answer: Filters documents.

Question 22: Securing MongoDB.

Answer: Enable auth, TLS.

Question 23: WiredTiger engine.

Answer: High performance storage engine.

In [15]:
from pymongo import MongoClient
import pandas as pd

# The connection to MongoDB at "mongodb://localhost:27017/" failed because a MongoDB server
# is not running or accessible in this Colab environment.
# To proceed with MongoDB operations, you would need to either:
# 1. Start a MongoDB server (e.g., using Docker or a local installation if outside Colab).
# 2. Connect to a cloud-hosted MongoDB service like MongoDB Atlas, providing its connection string.

# The following lines related to MongoDB connection and data insertion have been commented out
# to resolve the 'Connection refused' error in this cell.
# Please note that subsequent cells attempting to use 'collection' will likely fail
# because the MongoDB collection object has not been successfully initialized.

# client = MongoClient("mongodb://localhost:27017/")
# db = client["superstore_db"]
# collection = db["orders"]

# Load CSV file (exact name from os.listdir) - 'df' is already available from kernel state.
# df = pd.read_csv('/content/superstore (3).csv', encoding='latin1')

# Insert to MongoDB - 'records' is already available from kernel state.
# records = df.to_dict(orient='records')
# collection.insert_many(records) # This line would cause an error as 'collection' is not defined.

print("MongoDB connection and data insertion lines have been commented out.")
print("To interact with MongoDB, please ensure a MongoDB server is running and accessible,")
print("then uncomment and adjust the connection string accordingly.")
print("The 'df' (Pandas DataFrame) and 'records' (list of dictionaries) variables are still available for data manipulation in Python.")

MongoDB connection and data insertion lines have been commented out.
To interact with MongoDB, please ensure a MongoDB server is running and accessible,
then uncomment and adjust the connection string accordingly.
The 'df' (Pandas DataFrame) and 'records' (list of dictionaries) variables are still available for data manipulation in Python.


In [16]:
import pandas as pd

# Load the CSV file into a pandas DataFrame
df = pd.read_csv('/content/superstore (3).csv', encoding='latin1')

print("CSV file 'superstore (3).csv' loaded successfully into a DataFrame named 'df'.")
print(f"Shape of the DataFrame: {df.shape}")
print("First 5 rows of the DataFrame:")
print(df.head())

CSV file 'superstore (3).csv' loaded successfully into a DataFrame named 'df'.
Shape of the DataFrame: (9994, 21)
First 5 rows of the DataFrame:
   Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell

In [12]:
import os
os.listdir('/content')


['.config', 'superstore (3).csv', 'sample_data']

# Retrieve all documents from the collection

In [18]:
# Retrieve all documents from the collection
# all_docs = list(collection.find())

# # Print documents
# for doc in all_docs:
#     print(doc)

print("MongoDB connection not established. Please connect to a MongoDB server to perform this operation.")

MongoDB connection not established. Please connect to a MongoDB server to perform this operation.


In [31]:
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi

# IMPORTANT: Replace <db_password> with your actual database user password.
uri = "mongodb+srv://khrishnarathore878_db_user:<db_password>@cluster0.unskfpv.mongodb.net/?appName=Cluster0"

try:
    # Create a new client and connect to the server
    client = MongoClient(uri, server_api=ServerApi('1'))

    # Send a ping to confirm a successful connection
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")

    # Now that 'client' is connected, you can define 'db' and 'collection'
    # Replace 'superstore_db' and 'orders' with your actual database and collection names if different.
    db = client["superstore_db"]
    collection = db["orders"]
    print("MongoDB client, database, and collection initialized.")

except Exception as e:
    print(f"Error connecting to MongoDB: {e}")
    print("Please ensure your connection string is correct, your IP address is whitelisted, and your database user credentials are valid.")


Error connecting to MongoDB: The DNS query name does not exist: _mongodb._tcp.cluster0.unskfpv.mongodb.net.
Please ensure your connection string is correct, your IP address is whitelisted, and your database user credentials are valid.


### Connecting to MongoDB Atlas

Paste your MongoDB Atlas connection string below, replacing `<username>` and `<password>` with your actual database user credentials. Remember to configure Network Access and create a Database User in your Atlas project first, as described in the previous message.

In [39]:
from pymongo import MongoClient

# IMPORTANT: Replace <username>, <password>, and the rest of the URI with your actual MongoDB Atlas connection string.
# Example: "mongodb+srv://myusername:mypassword@cluster0.abcde.mongodb.net/?retryWrites=true&w=majority"
mongo_uri = "mongodb+srv://<username>:<password>@cluster0.abcde.mongodb.net/myDatabase?retryWrites=true&w=majority"

try:
    client = MongoClient(mongo_uri)

    # Replace 'mydatabase' with your actual database name if you have one, or keep 'superstore_db'
    db = client["superstore_db"]

    # Replace 'mycollection' with your actual collection name, or keep 'orders'
    collection = db["orders"]

    print("Successfully connected to MongoDB Atlas!")

    # You can now try to perform a simple operation, e.g., list existing collections
    # print("Collections in the database:", db.list_collection_names())

except Exception as e:
    print(f"Error connecting to MongoDB Atlas: {e}")
    print("Please ensure your connection string is correct, your IP address is whitelisted, and your database user credentials are valid.")


Error connecting to MongoDB Atlas: The DNS query name does not exist: _mongodb._tcp.cluster0.abcde.mongodb.net.
Please ensure your connection string is correct, your IP address is whitelisted, and your database user credentials are valid.


**Reasoning**:
To display the first 10 records from the `df` DataFrame as instructed, I will use the `.head()` method with an argument of 10.



In [32]:
df.head(10)

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
5       6  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
6       7  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
7       8  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
8       9  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
9      10  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
5  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
6  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
7  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
8  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
9  Brosina Hoffman   Consumer  United States      Los Angeles  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   
5       90032    West  FUR-FU-10001487        Furniture  Furnishings   
6       90032    West  OFF-AR-10002833  Office Supplies          Art   
7       90032    West  TEC-PH-10002275       Technology       Phones   
8       90032    West  OFF-BI-10003910  Office Supplies      Binders   
9       90032    West  OFF-AP-10002892  Office Supplies   Appliances   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   
5  Eldon Expressions Wood and Plastic Desk Access...   48.8600         7   
6                                         Newell 322    7.2800         4   
7                     Mitel 5320 IP Phone VoIP phone  907.1520         6   
8  DXL Angle-View Binders with Locking Rings by S...   18.5040         3   
9                   Belkin F5C206VTEL 6 Outlet Surge  114.9000         5   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  
5      0.00   14.1694  
6      0.00    1.9656  
7      0.20   90.7152  
8      0.20    5.7825  
9      0.00   34.4700  

[10 rows x 21 columns]

## Count All Records

### Subtask:
Count the total number of records in the `df` DataFrame, similar to `collection.count_documents({})`.

**Reasoning**:
To count the total number of records in the `df` DataFrame, I will use `len(df)` which returns the number of rows in the DataFrame.



In [33]:
record_count = len(df)
print(f"Total number of records in the DataFrame: {record_count}")

Total number of records in the DataFrame: 9994


## Filter Records by Region (West)

### Subtask:
Filter the `df` DataFrame to show records where the 'Region' is 'West', similar to `find({'Region': 'West'})`.

**Reasoning**:
To filter the `df` DataFrame to show records where the 'Region' is 'West', I will use boolean indexing on the 'Region' column.



In [34]:
west_region_df = df[df['Region'] == 'West']
print(f"Number of records in the West region: {len(west_region_df)}")
print("First 5 records from the West region:")
print(west_region_df.head())

Number of records in the West region: 3203
First 5 records from the West region:
   Row ID        Order ID Order Date  Ship Date       Ship Mode Customer ID  \
2       3  CA-2016-138688  6/12/2016  6/16/2016    Second Class    DV-13045   
5       6  CA-2014-115812   6/9/2014  6/14/2014  Standard Class    BH-11710   
6       7  CA-2014-115812   6/9/2014  6/14/2014  Standard Class    BH-11710   
7       8  CA-2014-115812   6/9/2014  6/14/2014  Standard Class    BH-11710   
8       9  CA-2014-115812   6/9/2014  6/14/2014  Standard Class    BH-11710   

     Customer Name    Segment        Country         City  ... Postal Code  \
2  Darrin Van Huff  Corporate  United States  Los Angeles  ...       90036   
5  Brosina Hoffman   Consumer  United States  Los Angeles  ...       90032   
6  Brosina Hoffman   Consumer  United States  Los Angeles  ...       90032   
7  Brosina Hoffman   Consumer  United States  Los Angeles  ...       90032   
8  Brosina Hoffman   Consumer  United States  Los Ange

## Filter Records by Sales (> 500)

### Subtask:
Filter the `df` DataFrame to show records where 'Sales' are greater than 500, similar to `find({'Sales': {'$gt': 500}})`.

**Reasoning**:
To filter the `df` DataFrame to show records where 'Sales' are greater than 500, I will use boolean indexing on the 'Sales' column.



In [35]:
high_sales_df = df[df['Sales'] > 500]
print(f"Number of records with Sales > 500: {len(high_sales_df)}")
print("First 5 records with Sales > 500:")
print(high_sales_df.head())

Number of records with Sales > 500: 1162
First 5 records with Sales > 500:
    Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
1        2  CA-2016-152156   11/8/2016  11/11/2016    Second Class   
3        4  US-2015-108966  10/11/2015  10/18/2015  Standard Class   
7        8  CA-2014-115812    6/9/2014   6/14/2014  Standard Class   
10      11  CA-2014-115812    6/9/2014   6/14/2014  Standard Class   
11      12  CA-2014-115812    6/9/2014   6/14/2014  Standard Class   

   Customer ID    Customer Name   Segment        Country             City  \
1     CG-12520      Claire Gute  Consumer  United States        Henderson   
3     SO-20335   Sean O'Donnell  Consumer  United States  Fort Lauderdale   
7     BH-11710  Brosina Hoffman  Consumer  United States      Los Angeles   
10    BH-11710  Brosina Hoffman  Consumer  United States      Los Angeles   
11    BH-11710  Brosina Hoffman  Consumer  United States      Los Angeles   

    ... Postal Code  Region       Product

## Find Top 3 by Profit

### Subtask:
Sort the `df` DataFrame by 'Profit' in descending order and display the top 3 records, similar to `find().sort('Profit', -1).limit(3)`.

**Reasoning**:
To find the top 3 records by 'Profit' in descending order, I will sort the `df` DataFrame by the 'Profit' column in descending order using `.sort_values()` and then select the first 3 rows using `.head(3)`.



In [36]:
top_3_profit_df = df.sort_values(by='Profit', ascending=False).head(3)
print("Top 3 records by Profit:")
print(top_3_profit_df)

Top 3 records by Profit:
      Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
6826    6827  CA-2016-118689   10/2/2016   10/9/2016  Standard Class   
8153    8154  CA-2017-140151   3/23/2017   3/25/2017     First Class   
4190    4191  CA-2017-166709  11/17/2017  11/22/2017  Standard Class   

     Customer ID Customer Name    Segment        Country       City  ...  \
6826    TC-20980  Tamara Chand  Corporate  United States  Lafayette  ...   
8153    RB-19360  Raymond Buch   Consumer  United States    Seattle  ...   
4190    HL-15040  Hunter Lopez   Consumer  United States     Newark  ...   

     Postal Code   Region       Product ID    Category Sub-Category  \
6826       47905  Central  TEC-CO-10004722  Technology      Copiers   
8153       98115     West  TEC-CO-10004722  Technology      Copiers   
4190       19711     East  TEC-CO-10004722  Technology      Copiers   

                               Product Name     Sales  Quantity  Discount  \
6826  Canon imageCL

## Update Ship Mode (First Class to Premium Class)

### Subtask:
Update the 'Ship Mode' column in the `df` DataFrame, changing all occurrences of 'First Class' to 'Premium Class'.

**Reasoning**:
To update the 'Ship Mode' column, replacing 'First Class' with 'Premium Class', I will use the `.replace()` method on the 'Ship Mode' column of the `df` DataFrame. I will then display the value counts of the 'Ship Mode' column to confirm the update.



In [37]:
df['Ship Mode'] = df['Ship Mode'].replace('First Class', 'Premium Class')
print("'First Class' in 'Ship Mode' column updated to 'Premium Class'.")
print("Updated 'Ship Mode' value counts:")
print(df['Ship Mode'].value_counts())

'First Class' in 'Ship Mode' column updated to 'Premium Class'.
Updated 'Ship Mode' value counts:
Ship Mode
Standard Class    5968
Second Class      1945
Premium Class     1538
Same Day           543
Name: count, dtype: int64


```markdown
## Filter Records by Sales (< 50)

### Subtask:
Filter the `df` DataFrame to show records where 'Sales' are less than 50, similar to `collection.delete_many({'Sales': {'$lt': 50}})`. This will be done by filtering and displaying the records, not actually deleting them.
```

## Filter Records by Sales (< 50)

### Subtask:
Filter the `df` DataFrame to show records where 'Sales' are less than 50, similar to `collection.delete_many({'Sales': {'$lt': 50}})`. This will be done by filtering and displaying the records, not actually deleting them.

**Reasoning**:
To filter the `df` DataFrame to show records where 'Sales' are less than 50, I will use boolean indexing on the 'Sales' column. I will then print the number of filtered records and display the first 5 records to verify the operation.



In [38]:
low_sales_df = df[df['Sales'] < 50]
print(f"Number of records with Sales < 50: {len(low_sales_df)}")
print("First 5 records with Sales < 50:")
print(low_sales_df.head())

Number of records with Sales < 50: 4849
First 5 records with Sales < 50:
   Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
5       6  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
6       7  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   
8       9  CA-2014-115812    6/9/2014   6/14/2014  Standard Class    BH-11710   

     Customer Name    Segment        Country             City  ...  \
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
5  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
6  Brosina Hoffman   Consumer  United States      Los Angeles  ...   
8  Brosina Hoffman   Consumer  United States      Los Angeles  ...   

  Postal Code  Regi

## Aggregate Total Sales by Region

### Subtask:
Group the `df` DataFrame by 'Region' and calculate the sum of 'Sales' for each region, similar to `collection.aggregate([{ '$group': {'_id': '$Region', 'Total_Sales': {'$sum': '$Sales'}} }])`.

**Reasoning**:
To group the `df` DataFrame by 'Region' and calculate the sum of 'Sales' for each region, I will use the `.groupby()` method on the 'Region' column followed by `.sum()` on the 'Sales' column.



In [40]:
total_sales_by_region = df.groupby('Region')['Sales'].sum().reset_index()
print("Total Sales by Region:")
print(total_sales_by_region)

Total Sales by Region:
    Region        Sales
0  Central  501239.8908
1     East  678781.2400
2    South  391721.9050
3     West  725457.8245


## Get Distinct Ship Modes

### Subtask:
Retrieve all unique values from the 'Ship Mode' column in the `df` DataFrame, similar to `distinct('Ship Mode')`.


**Reasoning**:
To retrieve all unique values from the 'Ship Mode' column, I will use the `.unique()` method on that specific column of the `df` DataFrame.

**Reasoning**:
To retrieve and display the distinct ship modes, I will apply the `.unique()` method to the 'Ship Mode' column of the `df` DataFrame and print the result.



In [41]:
distinct_ship_modes = df['Ship Mode'].unique()
print("Distinct Ship Modes:", distinct_ship_modes)

Distinct Ship Modes: ['Second Class' 'Standard Class' 'Premium Class' 'Same Day']


**Reasoning**:
To group the `df` DataFrame by 'Category' and count the number of orders in each category, I will use the `.groupby()` method on the 'Category' column followed by `.size()` to count the records in each group. `reset_index()` will convert the result back to a DataFrame with a proper column name for the count.



In [42]:
total_orders_by_category = df.groupby('Category').size().reset_index(name='Total_Orders')
print("Total Orders by Category:")
print(total_orders_by_category)

Total Orders by Category:
          Category  Total_Orders
0        Furniture          2121
1  Office Supplies          6026
2       Technology          1847


## Summary of Pandas Operations

All data manipulation tasks have been successfully performed using the `df` Pandas DataFrame, mimicking MongoDB operations without requiring an active MongoDB connection. Below is a summary of the results obtained:

1.  **First 10 Records**: Displayed the first 10 rows of the `df` DataFrame.
2.  **Total Record Count**: Found the total number of records to be `9994`.
3.  **Records in 'West' Region**: Identified `3203` records where the 'Region' is 'West'.
4.  **Records with Sales > 500**: Identified `1162` records where 'Sales' are greater than 500.
5.  **Top 3 by Profit**: Displayed the top 3 records with the highest 'Profit'.
6.  **Updated 'Ship Mode'**: Successfully changed 'First Class' to 'Premium Class'. The updated counts are: `Standard Class: 5968`, `Second Class: 1945`, `Premium Class: 1538`, `Same Day: 543`.
7.  **Records with Sales < 50**: Identified `4849` records where 'Sales' are less than 50.
8.  **Total Sales by Region**: Aggregated sales for each region:
    -   Central: `501239.8908`
    -   East: `678781.2400`
    -   South: `391721.9050`
    -   West: `725457.8245`
9.  **Distinct Ship Modes**: Retrieved `['Second Class' 'Standard Class' 'Premium Class' 'Same Day']`.
10. **Total Orders by Category**: Aggregated order counts by category:
    -   Furniture: `2121`
    -   Office Supplies: `6026`
    -   Technology: `1847`

These steps demonstrate the flexibility and power of Pandas for data analysis and manipulation, even when a direct database connection is not available.